transports/quic/src/muxer.rs

```diff
@@ -0,0 +1,360 @@
+            }
+        };
+        inner.substreams.insert(substream_id, Default::default());
+        let substream = Substream::new(substream_id, self.inner.clone());
+        Poll::Ready(Ok(substream))
+    }
+
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        let Inner {
+            substreams,
+            connection,
+            ..
+        } = &mut *self.inner.lock();
+        if connection.is_drained() {
+            return Poll::Ready(Ok(()));
+        }
+
+        if connection.send_stream_count() != 0 {
+            for substream in substreams.keys() {
+                if let Err(e) = connection.finish_substream(*substream) {
+                    log::warn!("substream finish error on muxer close: {}", e);
+                }
+            }
+        }
+        loop {
+            if connection.send_stream_count() == 0 && !connection.is_closed() {
+                connection.close()
+            }
+            match connection.poll_event(cx) {
+                Poll::Ready(ConnectionEvent::ConnectionLost(_)) => return Poll::Ready(Ok(())),
+                Poll::Ready(_) => {}
+                Poll::Pending => break,
+            }
+        }
+        Poll::Pending
+    }
+}
+
+pub struct Substream {
+    id: quinn_proto::StreamId,
+    muxer: Arc<Mutex<Inner>>,
+}
+
+impl Substream {
+    fn new(id: quinn_proto::StreamId, muxer: Arc<Mutex<Inner>>) -> Self {
+        Self { id, muxer }
+    }
+}
+
+impl AsyncRead for Substream {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        mut buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        let mut muxer = self.muxer.lock();
+
+        let mut stream = muxer.connection.recv_stream(self.id);
+        let mut chunks = match stream.read(true) {
+            Ok(chunks) => chunks,
+            Err(quinn_proto::ReadableError::UnknownStream) => {
+                return Poll::Ready(Ok(0));
+            }
+            Err(quinn_proto::ReadableError::IllegalOrderedRead) => {
+                unreachable!(
+                    "Illegal ordered read can only happen if `stream.read(false)` is used."
+                );
+            }
+        };
+        let mut bytes = 0;
+        let mut pending = false;
+        loop {
+            if buf.is_empty() {
+                break;
+            }
+            match chunks.next(buf.len()) {
+                Ok(Some(chunk)) => {
+                    buf.write_all(&chunk.bytes).expect("enough buffer space");
+                    bytes += chunk.bytes.len();
+                }
+                Ok(None) => break,
+                Err(err @ quinn_proto::ReadError::Reset(_)) => {
+                    return Poll::Ready(Err(io::Error::new(io::ErrorKind::ConnectionReset, err)))
+                }
+                Err(quinn_proto::ReadError::Blocked) => {
+                    pending = true;
+                    break;
+                }
+            }
+        }
```

Comment:
Could use early return to visually separate happy and unhappy paths.

```suggestion
        loop {
            if buf.is_empty() {
                break;
            }
            let chunk = match chunks.next(buf.len()) {
                Ok(Some(chunk)) => chunk,
                Ok(None) => break,
                Err(err @ quinn_proto::ReadError::Reset(_)) => {
                    return Poll::Ready(Err(io::Error::new(io::ErrorKind::ConnectionReset, err)))
                }
                Err(quinn_proto::ReadError::Blocked) => {
                    pending = true;
                    break;
                }
            };

            buf.write_all(&chunk.bytes).expect("enough buffer space");
            bytes += chunk.bytes.len();
        }
```